File: db4on/reference/vb/db4o-reference-chapters/src/com/db4odoc/f1/diagnostics/DiagnosticExample.vb
' Copyright (C) 2004 - 2006 db4objects Inc. http://www.db4o.com 

Imports com.db4o
Imports com.db4o.query
Imports com.db4o.diagnostic
Imports System
Imports System.IO

Namespace com.db4odoc.f1.diagnostics
    Public Class DiagnosticExample
        Inherits Util
        Public Shared Sub TestEmpty()
            Db4oFactory.Configure().Diagnostic().AddListener(New DiagnosticToConsole())
            File.Delete(Util.YapFileName)
            Dim db As ObjectContainer = Db4oFactory.OpenFile(Util.YapFileName)
            Try
                SetEmptyObject(db)
            Finally
                db.Close()
            End Try
        End Sub

        Private Shared Sub SetEmptyObject(ByVal db As ObjectContainer)
            Dim empty As Empty = New Empty()
            db.Set(empty)
        End Sub

        Public Shared Sub TestArbitrary()
            Db4oFactory.Configure().Diagnostic().AddListener(New DiagnosticToConsole())
            File.Delete(Util.YapFileName)
            Dim db As ObjectContainer = Db4oFactory.OpenFile(Util.YapFileName)
            Try
                Dim pilot As evaluations.Pilot = New evaluations.Pilot("Rubens Barrichello", 99)
                db.Set(pilot)
                QueryPilot(db)
            Finally
                db.Close()
            End Try
        End Sub

        Private Shared Sub QueryPilot(ByVal db As ObjectContainer)
            Dim i() As Integer = New Integer() {19, 100}

            Dim result As ObjectSet = db.Query(New ArbitraryQuery(i))
            ListResult(result)
        End Sub

        Public Shared Sub TestIndexDiagnostics()
            Db4oFactory.Configure().Diagnostic().RemoveAllListeners()
            Db4oFactory.Configure().Diagnostic().AddListener(New IndexDiagListener())
            Db4oFactory.Configure().UpdateDepth(3)
            File.Delete(Util.YapFileName)
            Dim db As ObjectContainer = Db4oFactory.OpenFile(Util.YapFileName)
            Try
                Dim pilot1 As evaluations.Pilot = New evaluations.Pilot("Rubens Barrichello", 99)
                db.Set(pilot1)
                Dim pilot2 As evaluations.Pilot = New evaluations.Pilot("Michael Schumacher", 100)
                db.Set(pilot2)
                QueryPilot(db)
                SetEmptyObject(db)
                Dim query As Query = db.Query()
                query.Constrain(GetType(evaluations.Pilot))
                query.Descend("_points").Constrain("99")
                Dim result As ObjectSet = query.Execute()
                ListResult(result)
            Finally
                db.Close()
            End Try
        End Sub

        Public Shared Sub TestTranslatorDiagnostics()
            StoreTranslatedCars()
            RetrieveTranslatedCars()
            RetrieveTranslatedCarsNQ()
            RetrieveTranslatedCarsNQUnopt()
            RetrieveTranslatedCarsSODAEv()
        End Sub

        Public Shared Sub StoreTranslatedCars()
            Db4oFactory.Configure().ExceptionsOnNotStorable(True)
            Db4oFactory.Configure().ObjectClass(GetType(evaluations.Car)).Translate(New evaluations.CarTranslator())
            Db4oFactory.Configure().ObjectClass(GetType(evaluations.Car)).CallConstructor(True)
            File.Delete(Util.YapFileName)
            Dim db As ObjectContainer = Db4oFactory.OpenFile(YapFileName)
            Try
                Dim car1 As evaluations.Car = New evaluations.Car("BMW")
                System.Diagnostics.Trace.WriteLine("ORIGINAL: " + car1.ToString())
                db.Set(car1)
                Dim car2 As evaluations.Car = New evaluations.Car("Ferrari")
                System.Diagnostics.Trace.WriteLine("ORIGINAL: " + car2.ToString())
                db.Set(car2)
            Catch exc As Exception
                System.Diagnostics.Trace.WriteLine(exc.Message)
                Return
            Finally
                db.Close()
            End Try
        End Sub

        Public Shared Sub RetrieveTranslatedCars()
            Db4oFactory.Configure().Diagnostic().RemoveAllListeners()
            Db4oFactory.Configure().Diagnostic().AddListener(New TranslatorDiagListener())
            Db4oFactory.Configure().ExceptionsOnNotStorable(True)
            Db4oFactory.Configure().ObjectClass(GetType(evaluations.Car)).Translate(New evaluations.CarTranslator())
            Db4oFactory.Configure().ObjectClass(GetType(evaluations.Car)).CallConstructor(True)
            Dim db As ObjectContainer = Db4oFactory.OpenFile(YapFileName)
            Try
                Dim query As Query = db.Query()
                query.Constrain(GetType(evaluations.Car))
                Dim result As ObjectSet = query.Execute()
                ListResult(result)
            Finally
                db.Close()
            End Try
        End Sub

        Public Shared Sub RetrieveTranslatedCarsNQ()
            Db4oFactory.Configure().Diagnostic().RemoveAllListeners()
            Db4oFactory.Configure().Diagnostic().AddListener(New TranslatorDiagListener())
            Db4oFactory.Configure().ExceptionsOnNotStorable(True)
            Db4oFactory.Configure().ObjectClass(GetType(evaluations.Car)).Translate(New evaluations.CarTranslator())
            Db4oFactory.Configure().ObjectClass(GetType(evaluations.Car)).CallConstructor(True)
            Dim db As ObjectContainer = Db4oFactory.OpenFile(YapFileName)
            Try
                Dim result As ObjectSet = db.Query(New NewCarModel())
                ListResult(result)
            Finally
                db.Close()
            End Try
        End Sub

        Public Shared Sub RetrieveTranslatedCarsNQUnopt()
            Db4oFactory.Configure().OptimizeNativeQueries(False)
            Db4oFactory.Configure().Diagnostic().RemoveAllListeners()
            Db4oFactory.Configure().Diagnostic().AddListener(New TranslatorDiagListener())
            Db4oFactory.Configure().ExceptionsOnNotStorable(True)
            Db4oFactory.Configure().ObjectClass(GetType(evaluations.Car)).Translate(New evaluations.CarTranslator())
            Db4oFactory.Configure().ObjectClass(GetType(evaluations.Car)).CallConstructor(True)
            Dim db As ObjectContainer = Db4oFactory.OpenFile(YapFileName)
            Try
                Dim result As ObjectSet = db.Query(New NewCarModel())
                ListResult(result)
            Finally
                Db4oFactory.Configure().OptimizeNativeQueries(True)
                db.Close()
            End Try
        End Sub

        Public Shared Sub RetrieveTranslatedCarsSODAEv()
            Db4oFactory.Configure().Diagnostic().RemoveAllListeners()
            Db4oFactory.Configure().Diagnostic().AddListener(New TranslatorDiagListener())
            Db4oFactory.Configure().ExceptionsOnNotStorable(True)
            Db4oFactory.Configure().ObjectClass(GetType(evaluations.Car)).Translate(New evaluations.CarTranslator())
            Db4oFactory.Configure().ObjectClass(GetType(evaluations.Car)).CallConstructor(True)
            Dim db As ObjectContainer = Db4oFactory.OpenFile(YapFileName)
            Try
                Dim query As Query = db.Query()
                query.Constrain(GetType(evaluations.Car))
                query.Constrain(New CarEvaluation())
                Dim result As ObjectSet = query.Execute()
                ListResult(result)
            Finally
                db.Close()
            End Try
        End Sub
    End Class

	
End Namespace

File: db4on/reference/db4o-reference-chapters/src/com/db4odoc/f1/diagnostics/DiagnosticExample.cs
/* Copyright (C) 2004 - 2006 db4objects Inc. http://www.db4o.com */

using com.db4o;
using com.db4o.query;
using com.db4o.diagnostic;
using com.db4odoc.f1.evaluations;
using System;
using System.IO;

namespace com.db4odoc.f1.diagnostics
{
	public class DiagnosticExample : Util
	{
		 public static void TestEmpty() {
    		Db4o.Configure().Diagnostic().AddListener(new DiagnosticToConsole());
			File.Delete(Util.YapFileName);    
			ObjectContainer db=Db4o.OpenFile(Util.YapFileName);
			try {
        		SetEmptyObject(db);
			}
			finally {
				db.Close();
			}
		}
	    
		private static void SetEmptyObject(ObjectContainer db){
    		Empty empty = new Empty();
			db.Set(empty);
		}
	    	
		public static void TestArbitrary() {
    		Db4o.Configure().Diagnostic().AddListener(new DiagnosticToConsole());
    		File.Delete(Util.YapFileName);    
			ObjectContainer db=Db4o.OpenFile(Util.YapFileName);
			try {
        		Pilot pilot = new Pilot("Rubens Barrichello",99);
        		db.Set(pilot);
        		QueryPilot(db);
			}
			finally {
				db.Close();
			}
		}
		
		private static void QueryPilot(ObjectContainer db){
    		int[]  i = new int[]{19,100};
    		ObjectSet result = db.Query(new ArbitraryQuery(i));
    		ListResult(result);
		}

		public static void TestIndexDiagnostics() {
    		Db4o.Configure().Diagnostic().RemoveAllListeners();
    		Db4o.Configure().Diagnostic().AddListener(new IndexDiagListener());
    		Db4o.Configure().UpdateDepth(3);
			File.Delete(Util.YapFileName);    
			ObjectContainer db=Db4o.OpenFile(Util.YapFileName);
			try {
        		Pilot pilot1 = new Pilot("Rubens Barrichello",99);
        		db.Set(pilot1);
        		Pilot pilot2 = new Pilot("Michael Schumacher",100);
        		db.Set(pilot2);
        		QueryPilot(db);
        		SetEmptyObject(db);
        		Query query = db.Query();
        		query.Constrain(typeof(Pilot));
				query.Descend("_points").Constrain("99");
				ObjectSet  result = query.Execute();
				ListResult(result);
			}
			finally {
				db.Close();
			}
		}
	     
		public static void TestTranslatorDiagnostics() {
    		StoreTranslatedCars();
    		RetrieveTranslatedCars();
    		RetrieveTranslatedCarsNQ();
    		RetrieveTranslatedCarsNQUnopt();
    		RetrieveTranslatedCarsSODAEv();
		}
	    
		public static void StoreTranslatedCars() {
    		Db4o.Configure().ExceptionsOnNotStorable(true);
    		Db4o.Configure().ObjectClass(typeof(Car)).Translate(new CarTranslator());
    		Db4o.Configure().ObjectClass(typeof(Car)).CallConstructor(true);
    		File.Delete(Util.YapFileName);    
			ObjectContainer db = Db4o.OpenFile(YapFileName);
			try {
				Car car1 = new Car("BMW");
				System.Diagnostics.Trace.WriteLine("ORIGINAL: " + car1);
				db.Set(car1);
				Car car2 = new Car("Ferrari");
				System.Diagnostics.Trace.WriteLine("ORIGINAL: " + car2);
				db.Set(car2);
			} catch (Exception exc) {
				System.Diagnostics.Trace.WriteLine(exc.Message);
				return;
			} finally {
				db.Close();
			}
		}

		public static void RetrieveTranslatedCars() {
    		Db4o.Configure().Diagnostic().RemoveAllListeners();
    		Db4o.Configure().Diagnostic().AddListener(new TranslatorDiagListener());
    		Db4o.Configure().ExceptionsOnNotStorable(true);
    		Db4o.Configure().ObjectClass(typeof(Car)).Translate(new CarTranslator());
    		Db4o.Configure().ObjectClass(typeof(Car)).CallConstructor(true);
    		ObjectContainer db = Db4o.OpenFile(YapFileName);
			try {
				Query query = db.Query();
				query.Constrain(typeof(Car));
				ObjectSet  result = query.Execute();
				ListResult(result);
			} finally {
				db.Close();
			}
		}

		public static void RetrieveTranslatedCarsNQ() {
    		Db4o.Configure().Diagnostic().RemoveAllListeners();
    		Db4o.Configure().Diagnostic().AddListener(new TranslatorDiagListener());
    		Db4o.Configure().ExceptionsOnNotStorable(true);
    		Db4o.Configure().ObjectClass(typeof(Car)).Translate(new CarTranslator());
    		Db4o.Configure().ObjectClass(typeof(Car)).CallConstructor(true);
    		ObjectContainer db = Db4o.OpenFile(YapFileName);
			try {
				ObjectSet  result = db.Query(new NewCarModel());
				ListResult(result);
			} finally {
				db.Close();
			}
		}
	    
		public static void RetrieveTranslatedCarsNQUnopt() {
    		Db4o.Configure().OptimizeNativeQueries(false);
    		Db4o.Configure().Diagnostic().RemoveAllListeners();
    		Db4o.Configure().Diagnostic().AddListener(new TranslatorDiagListener());
    		Db4o.Configure().ExceptionsOnNotStorable(true);
    		Db4o.Configure().ObjectClass(typeof(Car)).Translate(new CarTranslator());
    		Db4o.Configure().ObjectClass(typeof(Car)).CallConstructor(true);
    		ObjectContainer db = Db4o.OpenFile(YapFileName);
			try {
				ObjectSet  result = db.Query(new NewCarModel());
				ListResult(result);
			} finally {
				Db4o.Configure().OptimizeNativeQueries(true);
				db.Close();
			}
		}

		public static void RetrieveTranslatedCarsSODAEv() {
    		Db4o.Configure().Diagnostic().RemoveAllListeners();
    		Db4o.Configure().Diagnostic().AddListener(new TranslatorDiagListener());
    		Db4o.Configure().ExceptionsOnNotStorable(true);
    		Db4o.Configure().ObjectClass(typeof(Car)).Translate(new CarTranslator());
    		Db4o.Configure().ObjectClass(typeof(Car)).CallConstructor(true);
    		ObjectContainer db = Db4o.OpenFile(YapFileName);
			try {
				Query query = db.Query();
				query.Constrain(typeof(Car));
				query.Constrain(new CarEvaluation());
				ObjectSet  result = query.Execute();
				ListResult(result);
			} finally {
				db.Close();
			}
		}
	}
}
